src/realm/deppart/byfield_gpu_impl.hpp

```diff
@@ -0,0 +1,261 @@
+#pragma once
+#include "realm/deppart/byfield.h"
+#include "realm/deppart/byfield_gpu_kernels.hpp"
+#include "realm/deppart/partitions_gpu_impl.hpp"
+#include <cub/cub.cuh>
+#include "realm/nvtx.h"
+
+namespace Realm {
+
+template <int N, typename T>
+struct RectFieldVolumeOp {
+  __device__ __forceinline__
+  size_t operator()(const Rect<N,T>& r) const {
+    return r.volume();
+  }
+};
+
+// everything in header form:
+template <int N, typename T, typename FT>
+void GPUByFieldMicroOp<N,T,FT>::gpu_populate_bitmasks()
+{
+
+    nvtx_range_push("cuda", "byfield_gpu_populate_bitmasks");
+
+    cudaStream_t stream;
+    CUDA_CHECK(cudaStreamCreate(&stream), stream);
+
+    Memory my_mem = field_data[0].inst.get_location();
+
+    nvtx_range_push("cuda", "build parent and instance entries");
+
+    std::vector<size_t> inst_offsets(field_data.size() + 1);
+    size_t inst_size = 0;
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      inst_offsets[i] = inst_size;
+      if (field_data[i].index_space.dense()) {
+        inst_size += 1;
+      } else {
+        // only call get_entries() once per source
+        inst_size += field_data[i].index_space.sparsity.impl()->get_entries().size();
+      }
+    }
+    // final end offset
+    inst_offsets[field_data.size()] = inst_size;
+
+    RegionInstance inst_entries_instance = this->realm_malloc(inst_size * sizeof(SparsityMapEntry<N,T>), my_mem);
+    SparsityMapEntry<N,T>* d_inst_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(inst_entries_instance, 0).base);
+
+    size_t inst_pos = 0;
+    for (size_t i = 0; i < field_data.size(); ++i) {
```

Comment:
I can't tell what's happening here but it seems like there should be a way to create an allocation then turn it into a single cudaMemcpy call.

Reply:
Yeah, I’m saying that you could probably write it to populate one buffer in host memory and then issue a single memcpy to the device.

Reply:
I still think this could be 1 cudaMemcpy.